Buncha touchups to add features

LiruMouse opened the pull request:
Ignore spaces in parsing d strings
Allow dice expressions to be returned without loss of precision (not just as ints)
Allow returning a string of the pre-calculation, post-rolling expression, so the user can know what the results were before the math.

Also Fix a bug in which the result of the first roll of a dice expression would replace all of that expression in the string.
For example
`3d3+4+5d5+3d3`
could become
`(1 + 2 +  2)+4+(2 + 5 + 4 + 1 + 1)+(1 + 2 + 2)`
but will now become
`(1 + 2 + 2)+4+(2 + 5 + 4 + 1 + 1)+(1 + 3 + 3)`
Note that this bug is fixed in the last commit, but I believe if you'd prefer to just have this fix you can cherry pick it cleanly.

LiruMouse committed:
Add functions for retrieving the unresolved results of rolls before the math is compiled

Shifts code around that was originally in RandomDice's int Roll(string roll)
Adds string RolledString(string roll)
Adds int Compiled(string rolled) for compiling the math expressions returned by RolledString

Changes Dice to an abstract class because int Roll(string roll) should always call Compiled(RollString(roll))
int Roll(string roll) still maintains the same behavior.

LiruMouse committed:
Provided dice functions should always call the abstract d.

LiruMouse committed:
Add Dice.CompileObj, in case the end user wants more precision.

LiruMouse committed:
Show all results of PartialRolls in the RolledString (grouped in parentheses)

If no number is specified before d, the user is asking for 1.
Allow any number of spaces between the d and the numbers around it.

Adds PartialRoll.multi_d() to return an IEnumerable of ints from a roll,
PartialRoll.d() now returns the multi_d().Sum()

LiruMouse committed:
Don't replace each instance of a roll string with the first result of its roll!

cidthecoatrack commented:
Hi Lirusaito!

I appreciate the pull request, but I have some questions:

1. Your bug fix for replacing multiple die rolls (your last commit) is good, but you added no tests to confirm the particular bug.  Please do so.
2. Your other 4 commits - I wonder what the idea of the functionality is.  It seems to complicate the dice quite a bit (namely in that you lose all benefits of the interface by having to make abstract classes) for what seems to be minimal gain (getting a calculator string?  Seeing a collection of numbers to sum?)  If you have particular examples where that functionality would be useful, I would love to discuss them.

Lirusaito commented:
I'll write a test for 1 soon, sure!


As for what I'm doing with the other 4:

I didn't see much purpose in every implementation of the interface doing
the same `d6() => d(6);` for each die, among other things. In short, some
things can just be set in stone for all implementations.

My purposes for returning the pre-calculation string is to show the user
what their individual dice have rolled. Here's an example from the bot I am
using this in:
2:26 PM Liru said: !roll (2d3+5)/6
2:26 PM 🎵Electra🎵 said: ((3 + 2)+5)/6 = 1.66666666666667.

On Thu, Feb 18, 2016 at 2:12 PM, Karl Speer <notifications@github.com>
wrote:

> Hi Lirusaito!
>
> I appreciate the pull request, but I have some questions:
>
>    1. Your bug fix for replacing multiple die rolls (your last commit) is
>    good, but you added no tests to confirm the particular bug. Please do so.
>    2. Your other 4 commits - I wonder what the idea of the functionality
>    is. It seems to complicate the dice quite a bit (namely in that you lose
>    all benefits of the interface by having to make abstract classes) for what
>    seems to be minimal gain (getting a calculator string? Seeing a collection
>    of numbers to sum?) If you have particular examples where that
>    functionality would be useful, I would love to discuss them.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/DnDGen/RollGen/pull/6#issuecomment-185867988>.
>

cidthecoatrack commented:
Cool, I like your reasoning for the new features.  In addition to the test for #1, can you also add tests around your new features?

And I thought of one additional feature that fits in line with what you have already done: we have a generic `object` return and an `int`.  Let's add 1 more method that returns type `T`, so the user can easily convert on the fly.

Lirusaito commented:
While I'm not very certain how your test library works, I copied previous
tests for dice expression strings and went from there... you can tell me
how I did, once I figure out commit messages I'll update this.

On Fri, Feb 19, 2016 at 7:35 PM, Karl Speer <notifications@github.com>
wrote:

> Cool, I like your reasoning for the new features. In addition to the test
> for #1 <https://github.com/DnDGen/RollGen/issues/1>, can you also add
> tests around your new features?
>
> And I thought of one additional feature that fits in line with what you
> have already done: we have a generic object return and an int. Let's add
> 1 more method that returns type T, so the user can easily convert on the
> fly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/DnDGen/RollGen/pull/6#issuecomment-186467074>.
>

Lirusaito commented:
Oh, do you want a separate method for T, or can I just make it T
Compiled<T> and have it be Compiled<int> when you want an int?

On Fri, Feb 19, 2016 at 8:34 PM, Inusaito Sayori <inusaito@gmail.com> wrote:

> While I'm not very certain how your test library works, I copied previous
> tests for dice expression strings and went from there... you can tell me
> how I did, once I figure out commit messages I'll update this.
>
> On Fri, Feb 19, 2016 at 7:35 PM, Karl Speer <notifications@github.com>
> wrote:
>
>> Cool, I like your reasoning for the new features. In addition to the test
>> for #1 <https://github.com/DnDGen/RollGen/issues/1>, can you also add
>> tests around your new features?
>>
>> And I thought of one additional feature that fits in line with what you
>> have already done: we have a generic object return and an int. Let's add
>> 1 more method that returns type T, so the user can easily convert on the
>> fly.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/DnDGen/RollGen/pull/6#issuecomment-186467074>.
>>
>
>

cidthecoatrack commented:
I was thinking a separate method for `T` - and `int` would just use that method.  For many use cases, dice roller will want an `int` in the end, so I still want that default method available without having the specify the `int` type.

As for the test framework, if you are using Visual Studio, the tests use NUnit.  The package is installed in the solution, but you will need the install the NUnit 3 Runner VS extension, so that the test explorer can see and run the NUnit tests.

Take your best stab at them, and once you've submitted those changes, I'll take a look and make any further needed changes - or I can collaborate with you in a "code review" to show you the changes I would make and why.  Whichever you prefer.

LiruMouse committed:
Add T Compiled<T> to complement int Compiled

The user can get the Compiled version of a rolled_string in any reasonable type they wish.

LiruMouse committed:
Add test RollMultipleSameRollString as per request.

LiruMouse committed:
Add tests for new features added in my previous commits

RollFromStringSpaces(): Make sure roll strings with spaces work
RollFromStringNoQuantity(): Make sure roll strings that don't specify quantity work the same as if quantity specified was 1

LiruMouse committed:
Convert.ToInt32 ignores spaces, no need to trim first.

Lirusaito commented:
Alright, look good now?

cidthecoatrack commented:
Check out [my cleanup commit](https://github.com/DnDGen/RollGen/commit/53a26439961c099caac6178b5035c65c62aed32f) and see if it satisfies you.  I did some code refactoring/cleanup and added some more tests.  If you agree with the changes I have made, I will merge your pull request into the master branch and publish a new release

Lirusaito commented:
I'm going to break this down line by line.
On the name changes, I don't mind.

```cs
var sumOfRollsInParantheses = string.Format("({0})", sumOfRolls);
```
Misspelling of Parenthesis. And why not use string interpolation?

```cs
roll = roll.Remove(matchIndex, matchValue.Length);
roll = roll.Insert(matchIndex, sumOfRollsInParantheses);
```
Does this perform as well as the
```cs
roll = $"{roll.Substring(0, i)}({sumOfRolls}){roll.Substring(i+m.Length)}"
```
would?

```cs
if (string.IsNullOrEmpty(sections[0]) == false)
```
It's impossible for it to be null, it matched the regex. That's why I just did the quick ternary to default to 1.

```cs
public int Roll(string roll) => Roll<int>(roll);
```
Hah, I had that at one point, but then you said to have both so I amended the commit.

...
catch (InvalidCastException)
```
You make a good point about just casting to the type it already is.
Shouldn't we let them handle the exception? Maybe we could put in documentation "Throws InvalidCastException."
You don't seem to have a Compile<T>, it'd be nice to have this be the lower body of the Roll<T> function and change Roll<T> to
```cs
        public T Roll<T>(string roll) => Compile<T>(RollString(roll));
```

All these next changes look great to many...
```cs
Assert.That(rollWithoutQuantity, Is.EqualTo(630));
```
The RollFromString test already assured this.

```cs
Assert.That(roll, Is.EqualTo("(1 + 2 + 3 + 4 + 5 + 6 + 7)%(8 + 9 + 10 + 11 + 12 + 13 + 14)"));
```
The purpose of this test was to make sure that the string didn't match the previous one, checking that it's equal to what it should be more than does this, I'm not sure why it cannot just be generalized as str[0] != str[1], though. Either works though!

```cs
public void GetRawCompiledValue()
```
This test would be better if we had Dice.Compile<T> as I mentioned earlier, then we could do
`dice.Compile<int>(roll)` instead of `Convert.ToInt32(roll)`

```cs
public void GetDecimalCompiledValue()
```
This is nearly the same as ComputeExpression.
On that note, maybe we should call it Compute, rather than Compile, I was just going by the terminology of the library you're using.

Everything else looks good, sorry for leaving out that strings test thing!

cidthecoatrack commented:
1. More than one Parenthesis are parentheses.  There is one on each side, so the plural. (I was an English major back in college.)
2. Certainly my modifications are more easily readable (I'm a big proponent for Clean Code, as described by Robert C. Martin, where code should literally read like well-written prose.)  However, I will add some additional tests to verify performance.  It should be the same.
3. Correct, `null` can't actually happen - however, I tend to avoid ternaries for debugging reasons.  In general, I try to follow that a single line of code should only ever do 1 thing - determine a boolean if, set something, etc.  This tends to shake out such that ternaries are generally avoided, parentheses are (almost) always never inside other parentheses, etc.  Basically, it's a style thing.
4. Throwing the exception instead of catching it would certainly be valid.  Do you feel that would be preferred over returning whatever the default of `T` is?
5. I realized that `Compile<T>` and `Roll<T>` would literally be the same thing - they both return the same thing and perform the same action.  It's purely a naming difference.  Did you have a specific example where `Compile<T>` might be preferred/do something slightly different than `Roll<T>`?
6. Duplicate asserts never hurt - help ensure consistency.  As long as the performance of the test suite isn't hit, it's not a problem.
7. In general (this is another Clean Code thing), test asserts should be as specific as humanly possible, to ensure the best correctness you can.  For instance, let's say that the function returned the first string correct, but the second string as empty (due to some terrible, terrible bug).  Your original test would pass because `string.Empty != "(1 + 2 + 3 + 4 + 5 + 6 + 7)"`, even though the functionality is in fact wrong.  Asserting against what the specific results should be for that test case helps prevent that kind of bug from slipping through.
8. This test is just to specifically test the `object` return - the `Convert` statements are only there for the test.  In production code, I would certainly just call `Roll<T>` (or `Compile<T>`, depending on number 4)
9. The test is very similar yes - but there is a difference.  The `ComputeExpression` test does not explicitly call the `Roll<T>` function, while this test does.  Functionally, we know it does the same thing, but unit tests should ideally be written with no assumptions of how the underlying code works.  That way, if we decide that `T Roll<T>` and `int Roll` should operate differently for some reason, the tests will still be valid and will catch those potential differences.
10. Nice thought on the `Compute` name.  When I was writing the tests, I made the mistake of trying to pass in strings with die rolls (`2d3`) into the `Compile` functions, which caused it to bomb.  I was trying to think of a method name that would make it clear that it should NOT accept die rolls, but couldn't think of anything - so I kept your `Compile` name.  Any ideas?  I feel that `Compute` might still have that problem.

Lirusaito commented:
1. Typo on my part, Parentheses was misspelled as Parantheses
3. Personally, I've always favored single line assignments, but it's fine to me either way. How about `if (roll.Length != 0)`, then?
4. Yeah, since we don't know if the default could be mistaken for a legitimate result.
5. What I mean by `compile<T>` is to have
```cs
        public T Compile<T>(string rolled)
        {
            if (rawRoll is T)
                return (T)rawRoll;
 
            try
            {
                return (T)Convert.ChangeType(rawRoll, typeof(T));
            }
            catch (InvalidCastException)
            {
                return default(T);
            }
         }
        public T Roll<T>(string roll) => Compile(RollString(roll));
```
This would allow the user to `Compile<T>` the result of `RollString(roll)` and be able to hold on to the string returned by RollString. For instance:
```cs
var rolled = dice.RollString("4d6+4d7/7+(4d6/2)");
message.Respond($"{rolled} = {dice.Compile<decimal>(rolled)}";
```
7. Fair enough.
9. Ah! Good call!
10. I started out thinking "DoMath" I really wish there was an unambiguous yet elegant word for this, but it's beyond me. Documentation lines could explain, though.

cidthecoatrack commented:
1. Typo fixed, thanks for finding that. (Now that English major comment makes me sound like an ass)
2. I've preferred the `string.IsNullOrEmpty`, as it's a little more verbose about what you're doing (checking for an empty string) as opposed to checking its length.
3. Change made.
4. Ah, I see what you mean.  Change made.  I also pulled a number of things up into the abstract class, since regardless of implementation, the functions need to interact the same way.
5. I went with the `Compute` name - and also threw in some argument checks.  If the string to be computed has matching rolls via the `rollRegex`, then we'll throw an argument exception stating that the roll cannot be computed.
6. Fun note: it turns out that the performance of our string replacements are negligible - the real performance kicker is with Albatross.  In the new commit, you'll see I've added some tests just to see how long certain rolls take to compute.  When Albatross has to add around 1000 individual numbers, it can take a while, but still less than a second (so acceptable in my book).  It becomes unusable (a minute for 1 roll!) once you're rolling `10000d1`.  Considering how these dice are typically used, I am going to cap my tests at `1000d1000`, and hopefully people aren't trying to roll 10,000 of anything at one time :stuck_out_tongue: 

Take a look at the newest commit and see what you think.

Lirusaito commented:
`2`. An empty string will have a length of zero, but if you prefer you could be explicit with
```cs
if (sections[0] == string.Empty)
```
`4`. Awesome.
A few things on the things you moved around
```cs
        protected Regex rollRegex;

        public Dice()
        {
            rollRegex = new Regex("\\d* *d *\\d+");
        }
```
You could avoid having the constructor by just doing
```cs
protected Regex rollRegex = new Regex("\\d* *d *\\d+");
```
`5`. Should compute be derived and not just part of the abstract? Is there another way it could be implemented? I guess another parser, but since Albatross is already a library dependency, would deriving to change parser be worthwhile?

It also might be better if instead of
```cs
            var unrolledDieRolls = rollRegex.Matches(rolled);

            if (unrolledDieRolls.Count > 0)
```
you went with
```cs
if (rollRegex.IsMatch(rolled))
```
`6`. I've done higher than that only in my testing (flailing on a numberpad gives interesting results.)

cidthecoatrack commented:
1. Yes, `string.Empty.length == 0`, but the `string.IsNullOrEmpty` check is just slightly more verbose and generally safer (we know null isn't possible here, but if it ever were, we would be safe).  So I will definitely prefer that check
2. Another Clean Code thing: never do construction outside of a constructor.  While you eliminate an explicit constructor method, you still are implicitly running construction code when the regex is newed up.  (Man, I don't know if I've ever had to type out the past tense of "new" as a verb before.  Chrome spellcheck is pissed.)  Better to make everything as explicit as possible - easier to understand, easier to debug, easier to maintain.
3. ... `Compute` (the raw one returning `object`) is derived...  I did that specifically because of the Albatross dependency.  Really, if we wanted to get super nit-picky, there should probably be another class here - an expression evaluator of some sort - that would encapsulate the Albatross dependency, entirely separate from the `Random` dependency.  Then the `Dice` would take in an `ExpressionEvaluator` as a dependency.  But I don't think we need to do that here.  I might put an issue in to do that refactor later.
4. ... Are you sure you were looking at the most recent commit?  Should be e25276e97de122cc9b1ef2644f008d66914e39be .  Because in that commit, I do check the count.
5. How high did you test?  And what was the average response time?  It's probably also important to note that my home laptop is only a mid-level one from 2010 and really is showing its age.  It was never really cut out for intense programming demands (was meant for college, and that's about it), so that might very well be affecting the performance.  The Azure server where dndgen.com is hosted certainly will not have that  issue - and I suspect your environment where you are using this package will also fare better.  I need to invest in a better, high-performance laptop to use for programming - it's just the issue of getting the moola.

Double-check the commit you were looking at.  If that looks good, and I've answered your questions/thoughts here, and you don't have any more, then everything should be good to go :)

cidthecoatrack commented:
Just as a reference, this is the coding style I reference: https://cleancoders.com/landing  Robert C. Martin (or Uncle Bob) does videos, has written books, does lectures - check him out.  This stuff is awesome.

Lirusaito commented:
`2`. I don't really see how that's part of class construction, but alright.
`3`. Yeah, that sorta refactor might be nice.
`4`. Yeah you check the count, I'm saying you shouldn't have it collect any more than the first, instead you should use:
```cs
if (rollRegex.IsMatch(rolled))
```
This only looks once, so it's more optimal for this use.
`5`. I think I cleaned up most of it, but here's the last large one I could find:
```
11:39 AM Liru said: !roll   9266    d  90210
11:39 AM Liru said: Welp, that was a bad idea
11:39 AM 🎵Electra🎵 said: Liru: Error: Command Error: Message must be 2000 characters or less.
Parameter name: text
```
So it all happened within the same minute. If you want, I could poke the api for the exact timestamps?

Hah, just noticed I'm writing this as I'm playing DnD. And yeah, when I have some spare time, I'll look at that since you speak so highly of it ^^.

cidthecoatrack commented:
1. It's implicit - that code would be run as the class is initialized/constructed, so it is in effect constructor logic.
2. It is now enhancement #7
3. Ah, I had your comment backward.  Yes, that is better.  Change made.
4. No need - again, I think that there won't be many instances where people are rolling ~10,000 of anything, so we should be fine.

Hopefully last commit: 3fd74a4dde4f4d378832c83229beebada4549b69 .  If you approve, I will merge into master

Lirusaito commented:
I just noticed you're using the match, in that case you could do:
```cs
Match match = r.Match(text);
if (match.Success)
{
    var message = string.Format("Cannot compute unrolled die roll {0}", match.Value);
```
That aside, it's all good.

cidthecoatrack commented:
Nice catch, made that change.  I will merge this into master.  I will release it as a new version (10.1) this evening.
